cmd/sync/azure.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/compute/mgmt/compute"
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/network/mgmt/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// AzureClient allows you to get the list of IP addresses of VirtualMachines of a VirtualMachine Scale Set. It implements the CloudProvider interface
+type AzureClient struct {
+	config      *azureConfig
+	vmssClient  compute.VirtualMachineScaleSetsClient
+	ifaceClient network.InterfacesClient
+	authorizer  autorest.Authorizer
+}
+
+// NewAzureClient creates an AzureClient
+func NewAzureClient(data []byte) (*AzureClient, error) {
+	azureClient := &AzureClient{}
+	cfg, err := azureClient.parseAzureConfig(data)
+	if err != nil {
+		return nil, fmt.Errorf("error validating config: %v", err)
+	}
+
+	azureClient.config = cfg
+
+	err = azureClient.configure()
+	if err != nil {
+		return nil, fmt.Errorf("error configuring Azure Client: %v", err)
+	}
+
+	return azureClient, nil
+}
+
+// parseAzureConfig parses and validates AzureClient config
+func (client *AzureClient) parseAzureConfig(data []byte) (*azureConfig, error) {
+	cfg := &azureConfig{}
+	err := yaml.Unmarshal(data, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateAzureConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// GetPrivateIPsForScalingGroup returns the list of IP addresses of instances of the Virtual Machine Scale Set
+func (client *AzureClient) GetPrivateIPsForScalingGroup(name string) ([]string, error) {
+	var ips []string
+
+	ctx := context.Background()
```

Comment:
Yeah `TODO` and `Background` are both empty context. I was thinking of extending this in the future with maybe a main context in the client and then use Context with Timeouts, what do you think ? 

I can change them to TODO to better express the idea of something missing, although again, they are literally the same (I used Background instead of TODO because some CI tools don't allow you to pass lint with this one, but it seems ours do, so we're good) :

Ref from (https://golang.org/src/context/context.go): 
```
var (
	background = new(emptyCtx)
	todo       = new(emptyCtx)
)

// Background returns a non-nil, empty Context. It is never canceled, has no
// values, and has no deadline. It is typically used by the main function,
// initialization, and tests, and as the top-level Context for incoming
// requests.
func Background() Context {
	return background
}

// TODO returns a non-nil, empty Context. Code should use context.TODO when
// it's unclear which Context to use or it is not yet available (because the
// surrounding function has not yet been extended to accept a Context
// parameter).
func TODO() Context {
	return todo
}
```